edx_arch_experiments/codejail_service/views.py

```diff
@@ -0,0 +1,153 @@
+                {'type': 'null'},
+            ],
+        },
+        'unsafely': {'type': 'boolean'},
+    },
+    'required': ['code'],
+}
+
+# A note on the authorization model used here:
+#
+# We really just need one service account to be able to call this, and
+# then also allow is_staff to call it for convenience and debugging
+# purposes.
+#
+# To do this "right", I'd probably have to create an empty abstract
+# model, create a permission on it, create a group, add the permission
+# to the group, and add the service account to the group. Then I could
+# check the calling user's has_perm. If I wanted to use bridgekeeper
+# (as we're trying to do more of) I might be able to give bridgekeeper
+# a `@blanket_rule` that checks membership in the group, then use
+# bridgekeeper here instead of checking permissions directly, but it's
+# possible this wouldn't work because bridgekeeper might require there
+# to be a model instance to pass in (and there wouldn't be one, since
+# it's just an abstract model.)
+#
+# But... given that the service account will be is_staff, and I'll be
+# opening this up to is_staff alongside the intended service account,
+# and this is already a hacky intermediate solution... we can just use
+# the `IsAdminUser` permission class and be done with it.
+
+
+@api_view(['POST'])
+@parser_classes([FormParser, MultiPartParser])
+@permission_classes([IsAdminUser])
+def code_exec_view_v0(request):
+    """
+    Executes code in a codejail sandbox for a remote caller.
+
+    This implements the API used by edxapp's xmodule.capa.safe_exec.remote_exec.
+    It accepts a POST of a form containing a `payload` value and zero or more
+    extra files.
+
+    The payload is JSON and contains the parameters to be sent to codejail's
+    safe_exec (aside from `extra_files`). See payload_schema for type information.
+
+    This API does not permit `unsafely=true`.
+    """
+    if not CODEJAIL_SERVICE_ENABLED.is_enabled():
+        return Response("Codejail service not enabled", status=500)
+
+    params_json = request.data['payload']
+    params = json.loads(params_json)
+    jsonschema.validate(params, payload_schema)
+
+    complete_code = params['code']  # includes standard prolog
+    input_globals_dict = params.get('globals_dict') or {}
+    python_path = params.get('python_path') or []
+    limit_overrides_context = params.get('limit_overrides_context')
+    slug = params.get('slug')
+    unsafely = params.get('unsafely')
+
+    extra_files = request.FILES
+
+    # There's a risk of getting into a loop if e.g. the CMS asks the
+    # LMS to run codejail executions on its behalf, and the LMS is
+    # *also* inadvertently configured to call the LMS (itself).
+    # There's no good reason to have a chain of >2 services passing
+    # codejail requests along, so only allow execution here if we
+    # aren't going to pass it along to someone else.
+    if getattr(settings, 'ENABLE_CODEJAIL_REST_SERVICE', False):
```

Comment:
If we move this up a few lines we can avoid doing unnecessary work

Reply:
Nice, will do.

I should probably also change it to log + return a 500 rather than throwing an exception.